frontend/workflows/projectCatalog/src/catalog.tsx

```diff
@@ -0,0 +1,105 @@
+import React from "react";
+import { Grid, Paper, TextField, Toast, Typography, useNavigate } from "@clutch-sh/core";
+import { Box, CircularProgress } from "@material-ui/core";
+import SearchIcon from "@material-ui/icons/Search";
+
+import type { WorkflowProps } from ".";
+import catalogReducer from "./catalog-reducer";
+import ProjectCard from "./project-card";
+import { addProject, getProjects, removeProject } from "./storage";
+import type { CatalogState } from "./types";
+
+const initialState: CatalogState = {
+  projects: [],
+  search: "",
+  isLoading: false,
+  isSearching: false,
+  error: undefined,
+};
+
+const Catalog: React.FC<WorkflowProps> = ({ heading }) => {
+  const navigate = useNavigate();
+  const [state, dispatch] = React.useReducer(catalogReducer, initialState);
+
+  const navigateToProject = project => {
+    navigate(`/catalog/${project.name}`);
+  };
+
+  const setError = err => dispatch({ type: "HYDRATE_ERROR", payload: { result: err.message } });
+
+  React.useEffect(() => {
+    dispatch({ type: "HYDRATE_START" });
+    getProjects(
+      projects => dispatch({ type: "HYDRATE_END", payload: { result: projects } }),
+      setError
+    );
+  }, []);
+
+  const triggerProjectAdd = () => {
+    dispatch({ type: "SEARCH_START" });
+    addProject(
+      state.search,
+      projects => {
+        dispatch({ type: "ADD_PROJECT", payload: { projects } });
+        dispatch({ type: "SEARCH_END" });
+      },
+      e => {
+        dispatch({ type: "SEARCH_END" });
+        setError(e);
+      }
+    );
+  };
```

Comment:
yeah looking at this snippet specifically it definitely makes sense to push the complexity into the reducer. IMO this will be prone to breakage because we have logic spread across reducer, this function, and `addProject`.

Reply:
Yep, there's also some logic right now where it refetches everything each time on removal etc that I want to refactor. This seemed like a good starting point to get it up and we can refactor it before merging completely.

Reply:
Added a TODO into the code